src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs

```diff
@@ -1848,16 +1856,16 @@ private static BigInteger Subtract(uint[]? leftBits, int leftSign, uint[]? right
                 return left._sign | right._sign;
             }
 
-            uint[] x = left.ToUInt32Array();
-            uint[] y = right.ToUInt32Array();
+            ReadOnlySpan<uint> x = left.ToUInt32Array();
+            ReadOnlySpan<uint> y = right.ToUInt32Array();
             uint[] z = new uint[Math.Max(x.Length, y.Length)];
             uint xExtend = (left._sign < 0) ? uint.MaxValue : 0;
             uint yExtend = (right._sign < 0) ? uint.MaxValue : 0;
 
             for (int i = 0; i < z.Length; i++)
             {
-                uint xu = (i < x.Length) ? x[i] : xExtend;
-                uint yu = (i < y.Length) ? y[i] : yExtend;
+                uint xu = (i < x.Length) ? Unsafe.Add(ref GetReference(x), i) : xExtend;
```

Comment:
```suggestion
                uint xu = ((uint)i < (uint)x.Length) ? x[i] : xExtend;
```
By using `uint`-casts the JIT should elide the bound-checks and you don't need `Unsafe.Add` here.
Same on other places.

Reply:
Are you sure? `x` and `y` are not arrays, so element access is actually call of indexer with hard-coded bound checks. I think explicit intentions is better than implicit, but more verbose, yes.

Reply:
Yes. 

> `x` and `y` are not arrays

But spans, and the JIT has the same / similar optimizations as on arrays here.

Reply:
Agree. But what about implicit vs explicit intentions?

Reply:
Same question for `Unsafe.Add` replacement in other places. It's not clear for me the way how to choose between JIT optimizations that work under the hood and implicit/invisible for developer and premature optimization. IMHO, `uint` cast is not obvious.

Reply:
The `uint`-trick is used quite often in this repo, so I'm tempted to say it's obvious.

With `Unsafe` you are on unsafe terrain, which needs extra scrutinies due to security. Relying on the JIT to remove the bound-check where the JIT can prove that they're not needed is the safer side, hence the way to go -- especially when the resulting codegen will be the same / there will be no perf difference, but you're "safe".

Reply:
> the JIT can prove

That's the problem. It's not always obvious whether the code can be recognized as some pattern by JIT. For instance,
```csharp
//BigIntegerCalculator.AddSub.cs
private static void Add(ReadOnlySpan<uint> left, uint right, ref uint bits)
{
  long digit = (long)GetReference(left) + right;
  bits = unchecked((uint)digit);
  long carry = digit >> 32;

  for (int i = 1; i < left.Length; i++)
  {
    digit = Unsafe.Add(ref GetReference(left), i) + carry;
    Unsafe.Add(ref bits, i) = unchecked((uint)digit);
    carry = digit >> 32;
  }
  Unsafe.Add(ref bits, left.Length) = (uint)carry;
}
```
Here I can replace `bits` argument with `Span`. But the first line of code has access to the first element of `bits`. In this case, bound check cannot be eliminated. The same for the last line. JIT doesn't know that `left.Length` is less than or equal to `bits.Length`, even if it is proved by assertions.

Reply:
Removing of all bounds checks should be a non-goal for this PR. We do like bounds checks for security reasons. We are ok with using unsafe constructs to remove bounds checks only there is a significant measurable performance benefit in doing so.

Reply:
We ultimately need a balance between code that is fast and code that is safe/understandable.
In cases where we can trivially improve the code quality without significantly regressing the perf, we should do so (e.g. using `uint`).

In the case of the above code, we expect that `left` has at least one element and `bits` is `left.Length + 1`. The bounds check should be elided by the `for` loop in general and the first case isn't really so special. That is, it could be:
```csharp
private static void Add(ReadOnlySpan<uint> left, uint right, Span<uint> bits)
{
  long carry = right;

  for (int i = 0; i < left.Length; i++)
  {
    digit = left[(uint)i] + carry;
    bits[(uint)i] = unchecked((uint)digit);
    carry = digit >> 32;
  }

  bits[(uint)left.Length] = (uint)carry;
}
```

Reply:
I believe there is also some trick we can do to get the JIT to elide the bounds checks on `bits`, but I don't recall what it is.
Maybe @GrabYourPitchforks remembers, as I believe he has needed similar.

Reply:
Why not just declare `i` as uint?